### PR TITLE
Lifecycle events

### PR DESCRIPTION
**Pull request by FPtje**
Alright, I'm really enthusiastic about this. This PR enables this example:

https://fptje.github.io/miso-jswidget-example/

(Repository here: https://github.com/FPtje/miso-jswidget-example)

That's a [flatpickr](https://flatpickr.js.org/) widget, acting like it's been written in Miso instead of Javascript. You can click on the dates in the calendar, and the rest of the Miso app will update. You can click the Miso buttons and the widget will update. And there's a `toggle calendar visibility` button to prove that it plays nicely with Miso's diffing algorithm.

This kind of stuff was technically *possible* before, but with this PR, we can be **good** at it. This could be a *huge* selling point! It opens Miso up to a whole world of widgets that have already been made by other people. I'm imagining people making libraries that just bring Javascript widgets into the Miso world. You'd be able to use them without worrying about JS interop.

See #222.

# Current status:
**Don't merge yet!**
- Isomorphism works! (see [here](https://github.com/FPtje/miso-isomorphic-example/tree/test-feat-lifecycles))
- I'd like some feedback on (amongst others) the approach. Is this a good way to implement this feature?
- Before merging I reckon the commits need to be squashed. For now the commits show how this PR progressed over time.

### CODE REVIEW

File: ghcjs-src/Miso/Html/Internal.hs [hunk 1]
```diff
@@ -263,6 +271,32 @@ onWithOptions options eventName Decoder{..} toAction =
...
+-- the DOM element that was created. It can use those to embed third party
+-- widgets in the element. The @sink@ can be used to create callbacks for
+-- event listeners, and DOM element to attach the widget to.
+onCreated
+  :: ((action -> IO ()) -> Object -> action)
+  -> Attribute action
+onCreated toAction =
+  Attribute $ \sink n -> do
+    onCreatedArr <- SomeJSArray <$> getProp "onCreated" n
+    cb <- jsval <$> asyncCallback1 (sink . toAction sink . Object)
```
Comment (basvandijk) [inline ghcjs-src/Miso/Html/Internal.hs]: @FPtje I see this callback (and also the `onDestroy` callback) is never released. This will cause a memory leak since all objects reachable from `toAction` and `sink` will be kept alive as long as the RTS is running.

The question is when to release the callback. Since the callbacks are only executed once we could just release them after they're called as in:

```javascript
for (var i in obj.onCreated){
  var cb = obj.onCreated[i];
  cb(obj.domRef);
  h$release(cb);
}
```

Would that work?

Comment (FPtje) [reply ghcjs-src/Miso/Html/Internal.hs]: Yes it would, I've delayed doing this to combine it with #363

**Comment by cocreature** [comment]
I actually implemented something very much like this at some point but ended up discarding it because it seemed to fragile. The problem is that it is really easy to get `miso` to reuse a DOM node which would result in the lifecycle hooks not being triggered.

Let me try to make this clear using a concrete example:
1. Your DOM currently looks something like this
```html
<button>Active third party library</button>
<div>Third party library not loaded</div>
```
2. Clicking on the button results in the third party lib (e.g. `flatpickr` being loaded). Now let’s assume the third-party lib element is also a `div` (might not be true for `flatpickr` but you should be able to modify this example). Afterwards the DOM looks like this
```html
<button>Deactivate third party library</button>
<div>Third party widget</div>
```
In this case, `miso` would simply reuse the existing DOM element (see https://github.com/FPtje/miso/blob/1dc5b88c16a9ca6a2c22e738bea7c96e56acdacb/jsbits/diff.js#L53) which will result in the lifecycle hooks __not__ being triggered.

Now you can definitely argue that this is an edge case that doesn’t matter in practise but I don’t think it’s that contrived of an example. At the very least it should be made very explicit in the docs when this can happen. The only actual solution to this problem that I could come up with was to require that nodes with lifecycle hooks have globally unique identifiers. Then you can distinguish if an existing DOM node with the same tag is an older version of the “same” node (i.e. has the same identifier) or if it just happens to have the same tag.

**Comment by FPtje** [comment]
@cocreature You actually make a very, very good point. Just now I was testing my local changes to get this to be compatible with isomorphism, and just added `onCreated` and `onDestroyed` events to my [miso-isomorphic-example](https://github.com/FPtje/miso-isomorphic-example)

For context, the isomorphic example has two routes: home being the counter "- <number> +" example, and flipped being the counter reversed: "+ <number> -".

I had added `onCreated` and `onDestroyed` events to both buttons in both routes. Switching from the home to flipped route just switched over some properties on the buttons. As a result, neither `onDestroyed` nor `onCreated` got called.

This is rather unexpected, but creating and destroying in this case is *genuinely* not what Miso is doing. If this behaviour is accepted, we should indeed document it clearly.

**Comment by cocreature** [comment]
> This is rather unexpected, but creating and destroying in this case is genuinely not what Miso is doing. If this behaviour is accepted, we should indeed document it clearly.

It’s not what `miso` is doing but if your usecase is integrating third party libraries (which seems like the primary usecase for this feature), you still want your lifecycle hooks to run in this case. So I’m not sure just documenting that it might not work is super useful.

**Comment by FPtje** [comment]
We could perform some kind of logic in the node diffing algorithm, perhaps throwing a destroy/create when certain aspects of a node change.

Perhaps virtual DOM nodes can be specifically marked for the specific purpose of embedding a widget. When that mark changes, throw destroyed, created  or both events. Then force the end user to only be able to listen to life cycle events of nodes that are marked as widget embedders.

**Comment by FPtje** [comment]
I've pushed a commit to the branch that resolves the issue, albeit not being perfect. I've replaced `onCreated` and `onDeleted` with a single function `lifeCycleEvents`, which forces its user to give some `uniqueId`. With that unique identifier Miso knows when to call the events, even when the tag doesn't change.

I've updated the jswidget example repo, but I've also pushed a branch to the isomorphic example:

https://github.com/FPtje/miso-isomorphic-example/tree/test-feat-lifecycles

Besides proving that isomorphism still works, the example reproduces the situation @cocreature described. The `+` and `-` buttons both listen to life cycle events, which they log to console. Going from `home` to `flipped` (or the other way around)  switches  them around, but since the tags (`button`) don't change, the DOM elements aren't actually removed and recreated. Despite this, the events are still called.

Now I'm not too happy about the unique id it requires, as just pushes a part of the problem to the user, who is unexpectedly tasked with coming up with some unique identifier for their widgets. That's really annoying and *really* difficult to explain.

As I'm typing this, I'm thinking of a better solution: use wrapped empty JS objects instead of `MisoString` for unique identification.

```haskell
module Miso.Html.Internal (
  ...
  LifeCycleIdentifier -- Note: No constructor export!
  createLifeCycleIdentifier
  ...
)
...
newtype LifeCycleIdentifier = LifeCycleIdentifier JSVal

createLifeCycleIdentifier :: IO LifeCycleIdentifier
createLifeCycleIdentifier = jsval <$> create
```

It would release the burden of having to think of a unique name, but people would still have to store the thing in the `Model` of their widgets.

What do you think?

**Comment by RobertFischer** [comment]
Preliminary Expectoration
=======================
I love the idea and I 100% support the concept behind the pull request. I, too, almost invented something much like this. :)

Here's some armchair ideation from the cheap seats. YMMV.

TL;DR
=========
Leverage [Key](https://hackage.haskell.org/package/miso-0.13.0.0/docs/Miso-Html.html#t:Key) concept already in the VDOM, either by extending the definition of `Key` or implementing another kind of `Key`. Have a `WidgetKey` that can either be generic or specified: something like `newtype WidgetKey = WidgetKey (Maybe Key)` would work. Other alternative below.

Thinking
=========

It seems to me that you don't need to distinguish all VDOM nodes -- mostly, it's sufficient to distinguish between widget vs. non-widget VDOM nodes. So tagging them somehow would be more than sufficient. It would be kludgey and JavaScript-tastic, but you could possibly even get away with just giving widget nodes some property with an esoteric, borderline legal name: `"!miso-widget-root"` or something.

I would imagine/expect that you would have the option to also give a widget instance a key, which would be directly tied to a particular instantiation of a widget, so that multiple instantiations of the widget would trigger all the lifecycle events. That helps distinguishing between different instantiation of the same widget at the same place, where you will almost certainly have the same structure but you will want the lifecycle events to all fire.  This could correspond to the value of the `"!miso-widget-root"` property, but then you'd have to special-case checking to see if that changed....

So, overall, it seems better if you could have a new kind of `Key` -- either by making `Key` into a variant type or by creating an entirely different thing (`WidgetKey`) that looks and acts strikingly like `Key`.

While I Am Just Throwing Out Ideas Willy Nilly...
========================================

If you want to go really nuts, you could go whole hog and change `Key` to be:
```haskell
data Key = Key MisoString | WidgetKey |  WidgetInstanceKey MisoString
```

Now you've got `Key` as an extensible expression of a node's purpose, possibly with an identifier. 

But once you've got that, conceptually, every node has a `Key` -- `Nothing` is just a different variant on the type. So I'd even go so crazy as to argue that the type of the key property on the node should not be `Maybe Key` but rather `Key`, with one of the variants being `GenericKey`, and that's what is assigned by default.  So the real defintion for `Key` that I would advocate is this:
```haskell
data Key = GenericKey | Key MisoString | WidgetKey | WidgetInstanceKey MisoString
```

**Comment by FPtje** [comment]
Wow, I love the enthusiasm! I like the idea of leveraging the concept of `Key` for this purpose, though I'm not sure there should be a distinction between a `WidgetKey` and a `WidgetInstanceKey`. `WidgetKey` would mark the node as being a widget. With that Miso knows enough to distinguish between a node that holds a widget and a node that does not, but as you point out, it doesn't help Miso distinguish between two nodes that hold different widgets. This may be hard to explain to the end user, since the explanation would depend on a deep understanding of Miso's diffing algorithm. 

A user's wrong choice can cause widgets to end up in places they shouldn't be, and I suspect it may be difficult to figure out that the choice of `WidgetKey` over `WidgetInstanceKey` is what causes these issues.

I definitely like the idea of leveraging `Key` though. It ties it back to concepts that already exist within Miso.

**Comment by FPtje** [comment]
The commit above implements the wrapped empty JS objects solution. It relieves the end user from having to think of a unique name. Not only that, the entire problem can be contained within the component that wraps the JS widget, see example [here](https://github.com/FPtje/miso-jswidget-example/blob/master/src/Flatpickr.hs#L42). Users of the wrapped widget don't have to bother with it at all.

Naming wise I called the driver here `LifeCycleKey`. It's not exactly what @RobertFischer describes, but I like the conceptual relation it has to existing keys.

What do you think?

**Comment by dmjio** [comment]
I'd envision that we could keep life cycle events as simple (and pure) as possible, something I had in mind was like this:

```haskell
view = button [ onCreated OnCreateSomeButton, id "some-button" ] []

update OnCreateSomeButton m = m <# do
   ele <- getElementById "some-button"
   initFlatPickr ele
   pure NoOp
```

- This would obviate the need for `DOMNode`

- Using `Key` for this might sound nice because they are supposed to be unique, but they're only required to be unique amongst siblings.

- Adding an explicit `lifecycleKey` to every node in the virtual DOM is expensive and unnecessary (since not all nodes use `onCreated` and `onDestroyed`).

- I'd like to avoid the word `"Widget"` if at all possible.

**Comment by cocreature** [comment]
So I think we can separate the discussion in two parts:

1. How do we indicate the identity of a node. As @dmjio pointed out, `Key` is not really appropriate for this. However I think the HTML `id_` attribute is also not appropriate since it should be possible to indicate that the underlying node has changed while keeping the `id_` the same (the `id_` might be used for CSS so you can’t just arbitrarily change it). So overall, I like the approach in this PR. While it’s not particularly nice to have to initialize the keys in `IO`, it gets the job done and prevents users from accidentally reusing ids.
2. Which API is provided to react to the creation and deletion of nodes (I’m going to focus on node creation here but the same holds for deletion). @FPtje’s provides a function `(action -> IO ()) -> DOMNode -> action` while @dmjio provides a `action -> Attribute action`. I think the difference between those two is rather small. You can already get access to some kind of `sink` function in update by just writing to a channel and having a subscription that reads from the channel. I’m going to argue that both of these APIs don’t provide sufficient potential for abstraction and we should instead provide `(action -> IO ()) -> DOMNode -> IO ()` with the important part being that we end in `IO ()`. This API is strictly more general than the other APIs. Let me provide an example for why I think this API is more appropriate:
  
Let’s start with `textarea_ [onChange OnChange] []`.  Now we want to replace the `textarea` by some rich text editor provided by a third party library (e.g. `monaco`). With the API in this PR, we would need to add a constructor to our action field, modify our `update` function to handle the node creation event, then we could initialize the editor by calling the code from the third party lib and register an `onChange` event. This is quite a heavy change to the code. And worse, there is no way for some kind of `miso-monaco` to wrap all of this in an API that let’s users of this library write code like `monaco [onChange OnChange] []`. So every application that uses this node will have to modify their `update` function to handle the setup case and register the events that they care about (the code called in this case could be shared). Personally, I definitely would prefer if using third-party nodes was as easy as using builtin DOM nodes and I could just replace `textarea_` by `monaco`!

So what’s the downside of this API? Defining the function is going to require `CPP` in most cases since the `IO` actions will call JS functions. Now that’s obviously not nice but as mentioned before, this can be extracted to libraries, e.g. the `miso-monaco` library is going to use `CPP` internally but users of this library get a nice API and don’t have to use `CPP`. In addition to that `miso` could also provide a `(action -> Attribute action) -> (action -> IO ()) -> DOMNode -> IO ()` function (which uses `CPP` internally) which allows recovering the API proposed by @dmjio without users having to use `CPP`.

So overall, I think the API provided in this PR is insufficient since it doesn’t allow providing the same interface to DOM nodes which uses lifecycle events and nodes that don’t.

**Comment by FPtje** [comment]
Thanks @dmjio and @cocreature for the feedback!

The `DOMNode` can indeed go in favour of `getElementById`. For `lifecycleKey` (or whatever its name would be), I agree with @cocreature though in that the `id` of an element is not an optimal substitute.

---

> Adding an explicit lifecycleKey to every node in the virtual DOM is expensive and unnecessary 

It is indeed unnecessary. I can remove the line that sets it to `null` without affecting the diffing algorithm in a major way by checking for `undefined` instead of `null` [here](https://github.com/FPtje/miso/blob/3f614f255961bb7ed92f0b24a292acf5e6287b34/jsbits/diff.js#L53). The check would still be necessary, though, for the reasons described above.

---

As for @cocreature's second point: that's a really interesting idea. The [Flatpickr](https://github.com/FPtje/miso-jswidget-example/blob/master/src/Flatpickr.hs) example, based on the current PR, defines a complete component. That includes a `Model`, `update` function and a `viewModel` function. It's the whole shebang. If your idea is indeed viable, that all could be reduced to just some view function:

```haskell
flatpickr_ 
    :: LifeCycleId {- to be discussed -} 
    -> [EventsToListenToOrSomething action] 
    -> [Attribute action] 
    -> View action
```

Where `EventsToListenToOrSomething` is some user defined data type representing the kinds of event listeners it supports for that particular widget.

With your proposal to provide `(action -> IO ()) -> DOMNode -> IO ()`, I reckon it should be possible to do that, and have widgets be created and destroyed at the appropriate time, but I wonder about updating the widget from the Miso world. In the flatpickr example you can press `Previous day` and `Next day` buttons, which update the flatpickr widget, keeping it in sync. What would be the best way to do that in your view?

It's possible to, again, build a component around it and building a solution much like the one I currently have, but that feels icky if we can get the rest reduced to just a view function.

**Comment by FPtje** [comment]
I've updated the PR with some things I've discussed with @dmjio:

- Add `&& n.key === c.key` to `diffVNodes` in first conditional branch
- In `else` case add correct logic for calling `onDestroyed` on old, and `onCreated` before new
- Remove `DOMNode`
- Change `lifeCycleEvents` to `onCreated :: action -> Attribute action`, `onDestroyed`

Also updated the isomorphic and jswidgets example. The way to listen to events still irks me, though. As @cocreature suggested I created a sub ([here](https://github.com/FPtje/miso-jswidget-example/blob/master/src/Main.hs#L65)), which just listens to a `TChan` and sinks the actions it receives. What bothers me about it is that I have to ask the parent of the component to create that mechanism. It's a leaky abstraction :(

**Comment by dmjio** [comment]
> The way to listen to events still irks me, though. As @cocreature suggested I created a sub (here), which just listens to a TChan and sinks the actions it receives. What bothers me about it is that I have to ask the parent of the component to create that mechanism. It's a leaky abstraction

I'm unsure why this would be necessary, why has event listening now changed?

**Comment by FPtje** [comment]
Because the `onCreated` doesn't pass the sink to the action anymore. Its type is now `onCreated :: action -> Attribute action` instead of `onCreated :: ((action -> IO ()) -> action) -> Attribute action`.

In the latter version I could use that sink in the callbacks to easily get actions back into the update function. It's a really ad hoc way of doing that, though, and it couples the life cycle events too tightly to js widgets. With the sink left out the signature is simple and clean again, but leaves an open question as to how to elegantly create callbacks in javascript that go back to the update function.

On the other hand, the `TChan` subscription that feeds everything back to the update function may be a pattern on its own in a way...

**Comment by basvandijk** [comment]
@FPtje here's an implementation of your idea that we discussed this morning: https://github.com/LumiGuide/miso/commit/0b2b69378e5579b323c67f8933747a153f02fb01

**Comment by basvandijk** [comment]
@FPtje and here's the proof that with `scheduleIOWithSink` we get rid of the leaky abstraction: https://github.com/basvandijk/miso-jswidget-example/commit/421b63f12fd9c14e5fc0ed588b265e4b08ae9b79.

File: ghcjs-src/Miso/Html/Internal.hs [hunk 2]
```diff
@@ -123,9 +127,13 @@ node ns tag key attrs kids = View $ \sink -> do
   cssObj <- jsval <$> create
   propsObj <- jsval <$> create
   eventObj <- jsval <$> create
+  onCreatedArr <- jsval <$> JSArray.create
```
[the remark this answers is not in the document]
Comment (FPtje) [reply ghcjs-src/Miso/Html/Internal.hs]: Much like you can have multiple `onClick` listeners to the same element, you can have multiple life cycle events per element. Take the following example:

```haskell
div_ [onCreate A, onCreate B] []
```

It's type correct, but what should this do? Only throw either of the two or both?

[the remark this answers is not in the document]
Comment (dmjio) [reply ghcjs-src/Miso/Html/Internal.hs]: Trying to minimize allocations, this change will force an empty array on all DOM nodes, and I only anticipate a few nodes using `onCreated`, if ever. In this case both would get called. I think simply using `jsval <$> JSArray.create` would provide the same behavior if given `[onCreated A, onCreated B]`

[the remark this answers is not in the document]
Comment (dmjio) [reply ghcjs-src/Miso/Html/Internal.hs]: @FPtje, after rethinking, in the above scenario, `onCreated B` would overwrite `onCreated A`, only one would get called. Can we just use a single ref here instead? I don't think an array is necessary. But if it's too much work, this should be fine.

**Comment by smaccoun** [comment]
Do you think it would be possible to get an example of integrating a JS widget on the README? This is a really great feature

**Comment by TravisWhitaker** [comment]
Sorry for the necromancy, but are these uniqueness and node-reuse issues still relevant in Miso? If so, what exactly are the rules that must be obeyed?

Is there any documentation on the safe/correct way to integrate existing components?

**Comment by FPtje** [comment]
Yes, any node that uses onCreated or onDestroyed MUST have a `Key`.

**Comment by TravisWhitaker** [comment]
Specifically, it seems that in order to use `onCreated` and `onDestroyed`, the root node of the external component must be created with `nodeHtmlKeyed` with a unique `Key`. In my (extremely limited) testing it doesn't seem that the `Key` needs to appear in the DOM under the root node for correct operation.

Is this right? I'll make a little PR updating the docs if so.

**Comment by FPtje** [comment]
@TravisWhitaker yes, that is right. Coincidentally, a colleague noted that this requirement indeed wasn't documented last week, so I made a pr documenting the requirement: https://github.com/dmjio/miso/pull/482

In all fairness I should have documented this from the start. Sorry for the inconvenience.

**Comment by TravisWhitaker** [comment]
Ah sorry I missed that PR, thanks! In general, it seems that any node that happens to have a convenient unique index should be created with `nodeHtmlKeyed`, since it a (AFAIU) allows for faster diffing. There are `liKeyed_` and `trKeyed_` helpers for this, perhaps there should be more?

**Comment by FPtje** [comment]
It's slightly more nuanced. When using keys, _every_ node on a given level must have a key for the diffing algorithm to be faster. Take the following example:

```haskell
div_[]
  [ h1_ [] [text "MY LIST"]
  , li_ [] [text "foo"]
  , liKeyed_ "bar" [onCreated Bar] [text "bar"]
  ]
```
The `onCreated` event on the `liKeyed_` will work just fine, since the element with the attribute is keyed. However, the diffing algorithm will **not** use the key to make the diffing faster. For that to work, the `h1_` and the `li_` above `liKeyed_` must have a key as well. For the diffing algorithm to be fast, _every_ node at a given level must have a key.
